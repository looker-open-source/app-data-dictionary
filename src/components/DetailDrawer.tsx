/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { theme, Drawer, TableRow, TableDataCell } from '@looker/components'
import styled from 'styled-components'
import type {
  ILookmlModel,
  ILookmlModelExplore,
  ILookmlModelExploreField,
  IUser,
} from '@looker/sdk'
import { useHistory } from 'react-router'
import { DETAILS_PANE, COMMENTS_PANE } from '../utils/constants'
import { internalExploreURL, usePathNames } from '../utils/routes'
import type {
  ColumnDescriptor,
  FieldComments,
  CommentPermissions,
} from './interfaces'
import { FieldMetadata } from './FieldMetadata'

const TableRowCustom = styled(TableRow as any)`
  transition: background-color 0.3s ease;
  &.active,
  &:hover {
    background-color: ${theme.colors.ui1};
  }

  .disabled {
    visibility: hidden;
  }

  &:hover .disabled {
    visibility: visible;
  }

  td {
    word-break: break-word;
  }
  cursor: pointer;
`

export const DetailDrawer: React.FC<{
  columns: ColumnDescriptor[]
  explore: ILookmlModelExplore
  model: ILookmlModel
  field: ILookmlModelExploreField
  shownColumns: string[]
  tab: number
  setTab: (tabIndex: number) => void
  comments: string
  addComment: (newCommentStr: string, field: string) => void
  editComment: (newCommentStr: string, field: string) => void
  deleteComment: (newCommentStr: string, field: string) => void
  authors: IUser[]
  me: IUser
  permissions: CommentPermissions
}> = ({
  columns,
  explore,
  field,
  model,
  shownColumns,
  tab,
  setTab,
  comments,
  addComment,
  editComment,
  deleteComment,
  authors,
  me,
  permissions,
}) => {
  const history = useHistory()
  const path = usePathNames()

  const parsedComments = JSON.parse(comments)

  const getFieldCommentsLength = (field: string) => {
    const exploreComments = parsedComments[explore.name]
      ? parsedComments[explore.name]
      : {}
    const commentFields = Object.keys(exploreComments)
    if (commentFields.includes(field) && exploreComments[field].length > 0) {
      return exploreComments[field].length
    } else {
      return null
    }
  }

  const fieldComments =
    (parsedComments[explore.name] &&
      parsedComments[explore.name][field.name]) ||
    []
  const sortedComments = fieldComments.sort(
    (x: FieldComments, y: FieldComments) => {
      return x.timestamp - y.timestamp
    }
  )

  const canViewComments = () => {
    if (permissions.disabled) {
      return false
    } else {
      return (
        (permissions.reader && sortedComments.length > 0) ||
        permissions.writer ||
        permissions.manager
      )
    }
  }

  function paneUrl(pane: number) {
    field &&
      history.push(
        internalExploreURL({
          model: explore.model_name,
          explore: explore.name,
          field: field.name,
          tab: pane.toString(),
        })
      )
  }

  function closePaneUrl() {
    field &&
      history.push(
        internalExploreURL({
          model: explore.model_name,
          explore: explore.name,
        })
      )
  }

  function detailsPane() {
    paneUrl(DETAILS_PANE)
    setTab(DETAILS_PANE)
  }

  function commentsPane() {
    canViewComments() && paneUrl(COMMENTS_PANE)
    canViewComments() && setTab(COMMENTS_PANE)
  }

  return (
    <Drawer
      content={
        <FieldMetadata
          field={field}
          columns={columns}
          explore={explore}
          key={field.name}
          model={model}
          tab={path.detailPane ? parseInt(path.detailPane) : tab}
          detailsPane={detailsPane}
          commentsPane={commentsPane}
          sortedComments={sortedComments}
          addComment={addComment}
          editComment={editComment}
          deleteComment={deleteComment}
          fieldCommentLength={getFieldCommentsLength(field.name)}
          commentAuthors={authors}
          me={me}
          permissions={permissions}
          canViewComments={canViewComments()}
        />
      }
      isOpen={field.name === path.fieldName}
      onClose={closePaneUrl}
    >
      <TableRowCustom>
        {columns.map(({ Formatter, ...props }) => {
          if (shownColumns.includes(props.rowValueDescriptor)) {
            const unformattedValue =
              field[props.rowValueDescriptor as keyof ILookmlModelExploreField]
            const displayValue = unformattedValue
              ? String(unformattedValue)
              : undefined
            return (
              <TableDataCell
                color="text3"
                p="medium"
                pl="small"
                fontWeight="normal"
                key={props.rowValueDescriptor}
                maxWidth={props.maxWidth}
                minWidth={props.minWidth}
                onClick={
                  props.rowValueDescriptor === 'comment'
                    ? commentsPane
                    : detailsPane
                }
              >
                <Formatter
                  x={displayValue}
                  isRow={true}
                  field={field}
                  commentCount={getFieldCommentsLength(field.name)}
                  canComment={canViewComments()}
                  reader={permissions.reader}
                  tags={unformattedValue as string[]}
                />
              </TableDataCell>
            )
          }
          return undefined
        })}
      </TableRowCustom>
    </Drawer>
  )
}
