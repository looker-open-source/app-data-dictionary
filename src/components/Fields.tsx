/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import {
  Box,
  Flex,
  FlexItem,
  Heading,
  Table,
  TableHead,
  TableHeaderCell,
  TableRow,
  TableBody,
  theme,
} from '@looker/components'
import styled from 'styled-components'
import type {
  ILookmlModel,
  ILookmlModelExplore,
  ILookmlModelExploreField,
  IUser,
} from '@looker/sdk'
import { DETAILS_PANE } from '../utils/constants'
import type { ColumnDescriptor, CommentPermissions } from './interfaces'
import { DetailDrawer } from './DetailDrawer'

export const TableWrapper = styled(Box as any)`
  border-bottom: 0.5px solid ${theme.colors.ui2};

  &:last-child {
    border-bottom: none;
  }
`

// Sticky Table Header
export const StickyHeader = styled(TableHeaderCell as any)`
  @supports (position: sticky) {
    position: sticky;
    top: 0;
  }
`

export const Fields: React.FC<{
  columns: ColumnDescriptor[]
  explore: ILookmlModelExplore
  label: string
  model: ILookmlModel
  fields: ILookmlModelExploreField[]
  search: string
  shownColumns: string[]
  comments: string
  addComment: (newCommentStr: string, field: string) => void
  editComment: (newCommentStr: string, field: string) => void
  deleteComment: (newCommentStr: string, field: string) => void
  authors: IUser[]
  me: IUser
  permissions: CommentPermissions
}> = ({
  columns,
  explore,
  label,
  fields,
  model,
  search,
  shownColumns,
  comments,
  addComment,
  editComment,
  deleteComment,
  authors,
  me,
  permissions,
}) => {
  const [tab, setTab] = React.useState(DETAILS_PANE)
  return (
    <TableWrapper p="xxlarge">
      <Flex>
        <FlexItem>
          <Heading as="h2" fontWeight="semiBold" mb="large">
            {label}
          </Heading>
        </FlexItem>
      </Flex>
      <Flex flexDirection="column">
        <Table width="100%">
          <TableHead>
            <TableRow>
              {columns.map((column) => {
                if (shownColumns.includes(column.rowValueDescriptor)) {
                  return (
                    <StickyHeader
                      key={column.label}
                      backgroundColor="ui1"
                      fontWeight="medium"
                      color="text"
                      fontSize="small"
                      p="medium"
                      pl="small"
                    >
                      {column.label}
                    </StickyHeader>
                  )
                }
                return undefined
              })}
            </TableRow>
          </TableHead>
          <TableBody fontSize="small">
            {fields.map((field) => {
              if (
                !search ||
                (field.label_short &&
                  field.label_short
                    .toLowerCase()
                    .includes(search.toLowerCase())) ||
                (field.description &&
                  field.description
                    .toLowerCase()
                    .includes(search.toLowerCase())) ||
                (field.field_group_label &&
                  field.field_group_label
                    .toLowerCase()
                    .includes(search.toLowerCase()))
              ) {
                return (
                  <DetailDrawer
                    field={field}
                    columns={columns}
                    explore={explore}
                    key={field.name}
                    model={model}
                    shownColumns={shownColumns}
                    tab={tab}
                    setTab={setTab}
                    comments={comments}
                    addComment={addComment}
                    editComment={editComment}
                    deleteComment={deleteComment}
                    authors={authors}
                    me={me}
                    permissions={permissions}
                  />
                )
              }
              return undefined
            })}
          </TableBody>
        </Table>
      </Flex>
    </TableWrapper>
  )
}
