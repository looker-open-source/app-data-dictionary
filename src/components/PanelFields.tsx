/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { useState } from 'react'
import {
  Box,
  ButtonOutline,
  Flex,
  FlexItem,
  Heading,
  InputSearch,
  Paragraph,
  Spinner,
  theme,
} from '@looker/components'
import styled from 'styled-components'
import groupBy from 'lodash/groupBy'
import values from 'lodash/values'
import flatten from 'lodash/flatten'
import toPairs from 'lodash/toPairs'
import orderBy from 'lodash/orderBy'
import type {
  ILookmlModel,
  ILookmlModelExplore,
  ILookmlModelExploreField,
  IUser,
} from '@looker/sdk'
import humanize from 'humanize-string'
import { exploreURL } from '../utils/urls'
import { ExternalLink } from './ExternalLink'
import type { ColumnDescriptor, CommentPermissions } from './interfaces'
import { QuickSearch } from './QuickSearch'
import { ViewOptions } from './ViewOptions'
import { Fields } from './Fields'
import { DIMENSION, MEASURE } from './CategorizedLabel'

export const Main = styled(Box as any)`
  position: relative;
  width: 100%;
  min-height: 93vh;
`

const FullPage = styled(Box as any)`
  position: relative;
  display: flex;
  align-items: stretch;
  justify-content: center;
  align-items: center;
  width: 100%;
  min-height: 93vh;
  flex-direction: column;
`

const IntroText = styled(Paragraph as any)`
  text-align: center;
  margin-top: 5em;
  max-width: 40%;
  color: ${theme.colors.text1};
`

export const ExploreSearch = styled(InputSearch as any)`
  margin-top: 0;
`

export const defaultShowColumns = [
  'label_short',
  'description',
  'name',
  'type',
  'sql',
]

export const PanelFields: React.FC<{
  columns: ColumnDescriptor[]
  currentExplore: ILookmlModelExplore | null
  currentModel: ILookmlModel | null
  loadingExplore: string
  comments: string
  addComment: (newCommentStr: string, field: string) => void
  editComment: (newCommentStr: string, field: string) => void
  deleteComment: (newCommentStr: string, field: string) => void
  authors: IUser[]
  me: IUser
  permissions: CommentPermissions
}> = ({
  columns,
  currentExplore,
  currentModel,
  loadingExplore,
  comments,
  addComment,
  editComment,
  deleteComment,
  authors,
  me,
  permissions,
}) => {
  const [search, setSearch] = useState('')
  const [shownColumns, setShownColumns] = useState([
    ...columns
      .filter((d) => {
        return d.default
      })
      .map((d) => d.rowValueDescriptor),
  ])
  const [hasDescription, setHasDescription] = useState([])
  const [hasTags, setHasTags] = useState([])
  const [hasComments, setHasComments] = useState([])
  const [fieldTypes, setFieldTypes] = useState([])
  const [selectedFields, setSelectedFields] = useState([])

  const typeMaker = (field: ILookmlModelExploreField) => {
    return humanize(field.type.split('_')[0])
  }

  if (loadingExplore) {
    return (
      <Main p="xxlarge">
        <Flex alignItems="center" height="100%" justifyContent="center">
          <Spinner />
        </Flex>
      </Main>
    )
  }

  if (
    currentModel &&
    currentExplore &&
    currentModel.name === currentExplore.model_name
  ) {
    const fields = flatten(values(currentExplore.fields))
      .map((f) => typeMaker(f))
      .filter((value, index, self) => self.indexOf(value) === index)

    const revivedComments = JSON.parse(comments)
    const encExplores = Object.keys(revivedComments)
    const commentObj = encExplores.includes(currentExplore.name)
      ? revivedComments[currentExplore.name]
      : (revivedComments[currentExplore.name] = {})

    const allFilters = hasDescription.concat(
      hasTags,
      hasComments,
      fieldTypes,
      selectedFields
    )
    const groups = orderBy(
      toPairs(
        groupBy(
          flatten(values(currentExplore.fields)).filter((f) => {
            const commentFlag = !!(
              commentObj[f.name] && commentObj[f.name].length > 0
            )
            return (
              !f.hidden &&
              (allFilters.length === 0 ||
                ((hasDescription.length === 0 ||
                  (hasDescription.includes('yes') && f.description) ||
                  (hasDescription.includes('no') && !f.description)) &&
                  (hasTags.length === 0 ||
                    (hasTags.includes('yes') && f.tags.length > 0) ||
                    (hasTags.includes('no') && f.tags.length === 0)) &&
                  (hasComments.length === 0 ||
                    (hasComments.includes('yes') && commentFlag) ||
                    (hasComments.includes('no') && !commentFlag)) &&
                  (fieldTypes.length === 0 ||
                    (fieldTypes.includes('dimensions') &&
                      f.category === DIMENSION) ||
                    (fieldTypes.includes('measures') &&
                      f.category === MEASURE)) &&
                  (selectedFields.length === 0 ||
                    selectedFields.includes(typeMaker(f)))))
            )
          }),
          (f) => f.view_label
        )
      ),
      ([group]) => group
    )

    return (
      <Main pt="large">
        <Flex
          flexDirection="row"
          justifyContent="space-between"
          mt="large"
          mb="xxlarge"
          pl="xxlarge"
          pr="xxlarge"
        >
          <FlexItem>
            <Heading as="h1" fontWeight="semiBold">
              {currentModel.label}
            </Heading>
            <Heading as="h4" color="text1">
              Select a field for more information.
            </Heading>
          </FlexItem>
          <FlexItem>
            <ExternalLink target="_blank" href={exploreURL(currentExplore)}>
              <ButtonOutline mr="small">Explore</ButtonOutline>
            </ExternalLink>
            <ViewOptions
              columns={columns.filter((d) => {
                return d.rowValueDescriptor !== 'comment'
              })}
              shownColumns={shownColumns}
              setShownColumns={setShownColumns}
            />
          </FlexItem>
        </Flex>
        <Flex mt="xlarge" pl="xxlarge" pr="xxlarge">
          <FlexItem width="350px">
            <ExploreSearch
              hideSearchIcon
              placeholder="Filter fields in this Explore"
              mt="medium"
              onChange={setSearch}
              value={search}
            />
          </FlexItem>
        </Flex>

        <QuickSearch
          selectedFields={selectedFields}
          fields={fields}
          fieldTypes={fieldTypes}
          hasDescription={hasDescription}
          hasTags={hasTags}
          hasComments={hasComments}
          setSelectedFields={setSelectedFields}
          setFieldTypes={setFieldTypes}
          setHasDescription={setHasDescription}
          setHasTags={setHasTags}
          setHasComments={setHasComments}
          showComments={!permissions.disabled}
        />

        <Box>
          {groups.map((group) => {
            if (group[1].length > 0) {
              return (
                <Fields
                  columns={columns}
                  explore={currentExplore}
                  fields={group[1]}
                  key={group[0]}
                  label={group[0]}
                  model={currentModel}
                  search={search}
                  shownColumns={shownColumns}
                  comments={comments}
                  addComment={addComment}
                  editComment={editComment}
                  deleteComment={deleteComment}
                  authors={authors}
                  me={me}
                  permissions={permissions}
                />
              )
            }
            return undefined
          })}
        </Box>
      </Main>
    )
  } else {
    return (
      <FullPage>
        <div style={{ width: '30%' }}>
          <img
            src={
              'https://marketplace-api.looker.com/app-assets/data_dictionary_2x.png'
            }
            alt="Empty Image"
          />
        </div>
        <IntroText>
          Click on one of the Explores to the left to begin searching through
          your data. Youâ€™ll see labels, descriptions, SQL definitions, and more
          for each field.
        </IntroText>
      </FullPage>
    )
  }
}
