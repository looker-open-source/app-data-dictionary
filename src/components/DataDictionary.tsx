/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import {
  Chip,
  Flex,
  Heading,
  Spinner,
  ComponentsProvider,
  theme,
} from '@looker/components'
import humanize from 'humanize-string'
import styled from 'styled-components'
import { useAllModels, useComments } from '../utils/fetchers'
import './styles.css'
import { useCurrentModel, useCurrentExplore } from '../utils/routes'
import { PanelFields } from './PanelFields'
import SidebarToggle from './SidebarToggle'
import type {
  ColumnDescriptor,
  SidebarStyleProps,
  FormatterProps,
} from './interfaces'
import { SQLSnippet } from './SQLSnippet'
import { Sidebar } from './Sidebar'
import { CommentIcon } from './CommentIcon'

import { NoModelsAvailable } from './NoModelsAvailable'
import { CategorizedLabel } from './CategorizedLabel'

export const columns: ColumnDescriptor[] = [
  {
    name: 'comment-icon',
    label: ' ',
    rowValueDescriptor: 'comment',
    Formatter: function RenderCommentIcon({
      isRow,
      commentCount,
      canComment,
      reader,
    }: FormatterProps) {
      const showIcon = !reader || commentCount > 0
      if (canComment && isRow && showIcon) {
        return <CommentIcon count={commentCount} />
      } else {
        return null
      }
    },
    maxWidth: '4em',
    default: true,
  },
  {
    name: 'field-label',
    label: 'Field Label',
    rowValueDescriptor: 'label_short',
    Formatter: function RenderLabelShort({ x, field }: FormatterProps) {
      if (field.field_group_label && field.field_group_variant) {
        return <>{`${field.field_group_label} ${field.field_group_variant}`}</>
      } else {
        return <>{x}</>
      }
    },
    minWidth: '12em',
    default: true,
  },
  {
    name: 'category',
    label: 'Category',
    rowValueDescriptor: 'category',
    Formatter: function RenderCategoryLabel({ x, field }: FormatterProps) {
      return <CategorizedLabel label={x} category={field.category} />
    },
    minWidth: '10em',
    default: false,
  },
  {
    name: 'description',
    label: 'Description',
    rowValueDescriptor: 'description',
    Formatter: function RenderDescription({ x, isRow }: FormatterProps) {
      if (x && isRow && x.length > 200) {
        return <>{x.slice(0, 200) + '...'}</>
      }
      return <>{x}</>
    },
    minWidth: '10em',
    default: true,
  },
  {
    name: 'lookml-name',
    label: 'LookML Name',
    rowValueDescriptor: 'name',
    Formatter: function RenderLookmlName({ x }: FormatterProps) {
      return <>{x.replace(/\./g, '.\u200B')}</>
    },
    minWidth: '8em',
    default: true,
  },
  {
    name: 'type',
    label: 'Type',
    rowValueDescriptor: 'type',
    Formatter: function RenderType({ x }: FormatterProps) {
      return <>{humanize(x)}</>
    },
    minWidth: '8em',
    default: true,
  },
  {
    label: 'SQL',
    rowValueDescriptor: 'sql',
    Formatter: function RenderSqlSnippet({ x, isRow }: FormatterProps) {
      return <SQLSnippet isRow={isRow} src={x} />
    },
    minWidth: '10em',
    name: 'sql',
    default: true,
  },
  {
    name: 'tags',
    label: 'Tags',
    rowValueDescriptor: 'tags',
    Formatter: function RenderTags({ tags }: FormatterProps) {
      return (
        <>
          {tags &&
            tags.map((tag: string) => (
              <Chip disabled key={tag}>
                {tag}
              </Chip>
            ))}
        </>
      )
    },
    default: false,
  },
]

export const DataDictionary = () => {
  const unfilteredModels = useAllModels()
  const currentModel = useCurrentModel()
  const [sidebarOpen, setSidebarOpen] = React.useState(true)
  const [search, setSearch] = React.useState('')
  const { currentExplore, loadingExplore } = useCurrentExplore()
  const {
    comments,
    authors,
    me,
    permissions,
    addComment,
    editComment,
    deleteComment,
  } = useComments(currentExplore)

  let models

  if (unfilteredModels) {
    models = unfilteredModels.filter(
      (m) => m.explores && m.explores.some((e) => !e.hidden)
    )
    if (!models.length) {
      return <NoModelsAvailable />
    }
  }

  if (!models || !me) {
    return (
      <Flex alignItems="center" height="100vh" justifyContent="center">
        <Spinner />
      </Flex>
    )
  }

  const toggleFn = () => setSidebarOpen(!sidebarOpen)

  return (
    <ComponentsProvider>
      <div style={{ minWidth: '1200px' }}>
        <PageHeader>
          <Heading as="h1" fontSize="xlarge" fontWeight="semiBold">
            Data Dictionary
          </Heading>
        </PageHeader>
        <PageLayout open={sidebarOpen}>
          <LayoutSidebar>
            {sidebarOpen && (
              <Sidebar
                currentExplore={currentExplore}
                currentModel={currentModel}
                loadingExplore={loadingExplore}
                models={models}
                search={search}
                setSearch={setSearch}
              />
            )}
          </LayoutSidebar>
          <SidebarDivider open={sidebarOpen}>
            <SidebarToggle
              isOpen={sidebarOpen}
              onClick={toggleFn}
              headerHeight="114px"
            />
          </SidebarDivider>
          <PageContent>
            <PanelFields
              columns={columns}
              currentExplore={currentExplore}
              currentModel={currentModel}
              loadingExplore={loadingExplore}
              comments={comments}
              addComment={addComment}
              editComment={editComment}
              deleteComment={deleteComment}
              authors={authors}
              me={me}
              permissions={permissions}
            />
          </PageContent>
        </PageLayout>
      </div>
    </ComponentsProvider>
  )
}

// TODO fix this hard coded color as part of theming overhaul.
//      This was originally Looker purple with a strange background image.
//      Removing image resulted in Looker purple which is odd so hard
//      coding to Google blue. Still not great but better than before.
const PageHeader = styled(Flex as any)`
  background-color: #1a73e8;
  padding: ${theme.space.large};
  h1 {
    color: ${theme.colors.keyText};
  }
`

const PageLayout = styled.div<SidebarStyleProps>`
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: ${({ open }) =>
    open ? '16.625rem 0 1fr' : '1.5rem 0 1fr'};
  grid-template-areas: 'sidebar divider main';
  position: relative;
`

const PageContent = styled.div`
  grid-area: main;
  position: relative;
`

const LayoutSidebar = styled.aside`
  position: absolute;
  top: 0;
  bottom: 0;
  width: 16.625rem;
  grid-area: sidebar;
  z-index: 0;
`

const SidebarDivider = styled.div<SidebarStyleProps>`
  transition: border 0.3s;
  border-left: 1px solid
    ${({ theme, open }) => (open ? theme.colors.ui2 : 'transparent')};
  grid-area: divider;
  overflow: visible;
  position: relative;
  &:hover {
    border-left: 1px solid
      ${({ theme, open }) => (open ? theme.colors.ui2 : theme.colors.ui4)};
  }
`
