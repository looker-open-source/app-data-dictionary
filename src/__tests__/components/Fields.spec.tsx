/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import renderer from 'react-test-renderer'
import { theme } from '@looker/components'
import { ThemeProvider } from 'styled-components'
import type { IUser } from '@looker/sdk'
import { columns } from '../../components/DataDictionary'
import {
  mockGroups,
  mockCurrentExplore,
  mockCurrentModel,
} from '../MockData/MockData'
import { defaultShowColumns } from '../../components/PanelFields'

import { Fields } from '../../components/Fields'
import type { CommentPermissions } from '../../components/interfaces'

jest.mock('../../components/DetailDrawer', () => ({
  DetailDrawer: () => 'DetailDrawer',
}))

it('renders correctly', () => {
  const fn = jest.fn()
  const tree = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Fields
          columns={columns}
          explore={mockCurrentExplore}
          label={mockGroups[0][0]}
          model={mockCurrentModel}
          fields={mockGroups[0][1]}
          search={null}
          shownColumns={defaultShowColumns}
          comments=""
          addComment={fn}
          editComment={fn}
          deleteComment={fn}
          authors={[]}
          me={{} as IUser}
          permissions={{} as CommentPermissions}
        />
      </ThemeProvider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
