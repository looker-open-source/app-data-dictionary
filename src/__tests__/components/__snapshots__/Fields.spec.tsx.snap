// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="Box-sc-5738oh-0 sc-fqkvVR bXKNNZ iZTwfc"
>
  <div
    className="Flex-sc-1ak395a-0 cvboiO"
  >
    <div
      className="FlexItem-sc-1xhpm9o-0 hVeONI"
    >
      <h2
        className="TextBase-sc-90l5yt-0 Heading-sc-m5s1gl-0 fhYkSK bXhYYS"
      >
        Databases
      </h2>
    </div>
  </div>
  <div
    className="Flex-sc-1ak395a-0 ifrcWG"
  >
    <table
      className="Table-sc-1x77f3i-0 drJrhC"
    >
      <thead
        className="TableHead-sc-1bzogra-0 hbGfLO"
      >
        <tr
          className="TableRow-sc-1ty7jg8-0 gyYQVo"
        >
          <th
            className="TableHeaderCell-sc-hi6xhd-0 sc-dcJsrY dcGAwl ffcdhp"
          >
            Field Label
          </th>
          <th
            className="TableHeaderCell-sc-hi6xhd-0 sc-dcJsrY dcGAwl ffcdhp"
          >
            Description
          </th>
          <th
            className="TableHeaderCell-sc-hi6xhd-0 sc-dcJsrY dcGAwl ffcdhp"
          >
            LookML Name
          </th>
          <th
            className="TableHeaderCell-sc-hi6xhd-0 sc-dcJsrY dcGAwl ffcdhp"
          >
            Type
          </th>
          <th
            className="TableHeaderCell-sc-hi6xhd-0 sc-dcJsrY dcGAwl ffcdhp"
          >
            SQL
          </th>
        </tr>
      </thead>
      <tbody
        className="TableBody-sc-1c7fa9v-0 JFfMx"
      >
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
        DetailDrawer
      </tbody>
    </table>
  </div>
</div>
`;
